// jotai/esm/devtools.js
import { useRef, useEffect, useContext, useCallback } from 'react';
import { useAtom, SECRET_INTERNAL_getStoreContext, SECRET_INTERNAL_useMutableSource } from 'jotai';

function useAtomDevtools(anAtom, name) {
  let extension;
  try {
    extension = window.__REDUX_DEVTOOLS_EXTENSION__;
  } catch {
  }
  if (!extension) {
    if (typeof process === "object" && process.env.NODE_ENV === "development" && typeof window !== "undefined") {
      console.warn("Please install/enable Redux devtools extension");
    }
  }
  const [value, setValue] = useAtom(anAtom);
  const lastValue = useRef(value);
  const isTimeTraveling = useRef(false);
  const devtools = useRef();
  const atomName = name || anAtom.debugLabel || anAtom.toString();
  useEffect(() => {
    if (extension) {
      devtools.current = extension.connect({ name: atomName });
      const unsubscribe = devtools.current.subscribe((message) => {
        var _a, _b, _c, _d, _e, _f, _g;
        if (message.type === "DISPATCH" && message.state) {
          if (((_a = message.payload) == null ? void 0 : _a.type) === "JUMP_TO_ACTION" || ((_b = message.payload) == null ? void 0 : _b.type) === "JUMP_TO_STATE") {
            isTimeTraveling.current = true;
          }
          setValue(JSON.parse(message.state));
        } else if (message.type === "DISPATCH" && ((_c = message.payload) == null ? void 0 : _c.type) === "COMMIT") {
          (_d = devtools.current) == null ? void 0 : _d.init(lastValue.current);
        } else if (message.type === "DISPATCH" && ((_e = message.payload) == null ? void 0 : _e.type) === "IMPORT_STATE") {
          const actions = (_f = message.payload.nextLiftedState) == null ? void 0 : _f.actionsById;
          const computedStates = ((_g = message.payload.nextLiftedState) == null ? void 0 : _g.computedStates) || [];
          computedStates.forEach(({ state }, index) => {
            var _a2;
            actions[index] || atomName;
            if (index === 0) {
              (_a2 = devtools.current) == null ? void 0 : _a2.init(state);
            } else {
              setValue(state);
            }
          });
        }
      });
      devtools.current.shouldInit = true;
      return unsubscribe;
    }
  }, [anAtom, extension, atomName, setValue]);
  useEffect(() => {
    if (devtools.current) {
      lastValue.current = value;
      if (devtools.current.shouldInit) {
        devtools.current.init(value);
        devtools.current.shouldInit = false;
      } else if (isTimeTraveling.current) {
        isTimeTraveling.current = false;
      } else {
        devtools.current.send(`${atomName} - ${new Date().toLocaleString()}`, value);
      }
    }
  }, [anAtom, extension, atomName, value]);
}

var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));
const hasInitialValue = (atom) => "init" in atom;
const IS_EQUAL_PROMISE = Symbol();
const INTERRUPT_PROMISE = Symbol();
const isInterruptablePromise = (promise) => !!promise[INTERRUPT_PROMISE];
const createInterruptablePromise = (promise) => {
  let interrupt;
  const interruptablePromise = new Promise((resolve, reject) => {
    interrupt = resolve;
    promise.then(resolve, reject);
  });
  interruptablePromise[IS_EQUAL_PROMISE] = (p) => p === interruptablePromise || p === promise;
  interruptablePromise[INTERRUPT_PROMISE] = interrupt;
  return interruptablePromise;
};
const createState = (initialValues, stateListener) => {
  const state = {
    l: stateListener,
    v: 0,
    a: new WeakMap(),
    m: new WeakMap(),
    p: new Map()
  };
  if (initialValues) {
    for (const [atom, value] of initialValues) {
      const atomState = { v: value, r: 0, d: new Map() };
      if (typeof process === "object" && process.env.NODE_ENV !== "production") {
        Object.freeze(atomState);
        if (!hasInitialValue(atom)) {
          console.warn("Found initial value for derived atom which can cause unexpected behavior", atom);
        }
      }
      state.a.set(atom, atomState);
    }
  }
  return state;
};
const getAtomState = (state, atom) => state.a.get(atom);
const wipAtomState = (state, atom, dependencies) => {
  const atomState = getAtomState(state, atom);
  const nextAtomState = __spreadProps(__spreadValues({
    r: 0
  }, atomState), {
    d: dependencies ? new Map(Array.from(dependencies).map((a) => {
      var _a, _b;
      return [
        a,
        (_b = (_a = getAtomState(state, a)) == null ? void 0 : _a.r) != null ? _b : 0
      ];
    })) : (atomState == null ? void 0 : atomState.d) || new Map()
  });
  return [nextAtomState, (atomState == null ? void 0 : atomState.d) || new Map()];
};
const setAtomValue = (state, atom, value, dependencies, promise) => {
  var _a, _b;
  const [atomState, prevDependencies] = wipAtomState(state, atom, dependencies);
  if (promise && !((_a = atomState.p) == null ? void 0 : _a[IS_EQUAL_PROMISE](promise))) {
    return;
  }
  (_b = atomState.c) == null ? void 0 : _b.call(atomState);
  delete atomState.e;
  delete atomState.p;
  delete atomState.c;
  delete atomState.i;
  if (!("v" in atomState) || !Object.is(atomState.v, value)) {
    atomState.v = value;
    ++atomState.r;
  }
  commitAtomState(state, atom, atomState, dependencies && prevDependencies);
};
const setAtomReadError = (state, atom, error, dependencies, promise) => {
  var _a, _b;
  const [atomState, prevDependencies] = wipAtomState(state, atom, dependencies);
  if (promise && !((_a = atomState.p) == null ? void 0 : _a[IS_EQUAL_PROMISE](promise))) {
    return;
  }
  (_b = atomState.c) == null ? void 0 : _b.call(atomState);
  delete atomState.p;
  delete atomState.c;
  delete atomState.i;
  atomState.e = error;
  commitAtomState(state, atom, atomState, prevDependencies);
};
const setAtomReadPromise = (state, atom, promise, dependencies) => {
  var _a, _b;
  const [atomState, prevDependencies] = wipAtomState(state, atom, dependencies);
  if ((_a = atomState.p) == null ? void 0 : _a[IS_EQUAL_PROMISE](promise)) {
    return;
  }
  (_b = atomState.c) == null ? void 0 : _b.call(atomState);
  if (isInterruptablePromise(promise)) {
    atomState.p = promise;
    delete atomState.c;
  } else {
    const interruptablePromise = createInterruptablePromise(promise);
    atomState.p = interruptablePromise;
    atomState.c = interruptablePromise[INTERRUPT_PROMISE];
  }
  commitAtomState(state, atom, atomState, prevDependencies);
};
const setAtomInvalidated = (state, atom) => {
  const [atomState] = wipAtomState(state, atom);
  atomState.i = atomState.r;
  commitAtomState(state, atom, atomState);
};
const setAtomWritePromise = (state, atom, promise) => {
  const [atomState] = wipAtomState(state, atom);
  if (promise) {
    atomState.w = promise;
  } else {
    delete atomState.w;
  }
  commitAtomState(state, atom, atomState);
};
const scheduleReadAtomState = (state, atom, promise) => {
  promise.finally(() => {
    readAtomState(state, atom, true);
  });
};
const readAtomState = (state, atom, force) => {
  if (!force) {
    const atomState = getAtomState(state, atom);
    if (atomState) {
      atomState.d.forEach((_, a) => {
        if (a !== atom) {
          const aState = getAtomState(state, a);
          if (aState && !aState.e && !aState.p && aState.r === aState.i) {
            readAtomState(state, a, true);
          }
        }
      });
      if (Array.from(atomState.d.entries()).every(([a, r]) => {
        const aState = getAtomState(state, a);
        return aState && !aState.e && !aState.p && aState.r !== aState.i && aState.r === r;
      })) {
        return atomState;
      }
    }
  }
  let error;
  let promise;
  let value;
  const dependencies = new Set();
  try {
    const promiseOrValue = atom.read((a) => {
      dependencies.add(a);
      if (a !== atom) {
        const aState2 = readAtomState(state, a);
        if (aState2.e) {
          throw aState2.e;
        }
        if (aState2.p) {
          throw aState2.p;
        }
        return aState2.v;
      }
      const aState = getAtomState(state, a);
      if (aState) {
        if (aState.p) {
          throw aState.p;
        }
        return aState.v;
      }
      if (hasInitialValue(a)) {
        return a.init;
      }
      throw new Error("no atom init");
    });
    if (promiseOrValue instanceof Promise) {
      promise = promiseOrValue.then((value2) => {
        setAtomValue(state, atom, value2, dependencies, promise);
        flushPending(state);
      }).catch((e) => {
        if (e instanceof Promise) {
          scheduleReadAtomState(state, atom, e);
          return e;
        }
        setAtomReadError(state, atom, e instanceof Error ? e : new Error(e), dependencies, promise);
        flushPending(state);
      });
    } else {
      value = promiseOrValue;
    }
  } catch (errorOrPromise) {
    if (errorOrPromise instanceof Promise) {
      promise = errorOrPromise;
    } else if (errorOrPromise instanceof Error) {
      error = errorOrPromise;
    } else {
      error = new Error(errorOrPromise);
    }
  }
  if (error) {
    setAtomReadError(state, atom, error, dependencies);
  } else if (promise) {
    setAtomReadPromise(state, atom, promise, dependencies);
  } else {
    setAtomValue(state, atom, value, dependencies);
  }
  return getAtomState(state, atom);
};
const canUnmountAtom = (atom, mounted) => !mounted.l.size && (!mounted.d.size || mounted.d.size === 1 && mounted.d.has(atom));
const invalidateDependents = (state, atom) => {
  const mounted = state.m.get(atom);
  mounted == null ? void 0 : mounted.d.forEach((dependent) => {
    if (dependent === atom) {
      return;
    }
    setAtomInvalidated(state, dependent);
    invalidateDependents(state, dependent);
  });
};
const writeAtomState = (state, atom, update) => {
  var _a;
  const writePromise = (_a = getAtomState(state, atom)) == null ? void 0 : _a.w;
  if (writePromise) {
    writePromise.then(() => {
      writeAtomState(state, atom, update);
      flushPending(state);
    });
    return;
  }
  const promiseOrVoid = atom.write((a) => {
    const aState = readAtomState(state, a);
    if (aState.e) {
      throw aState.e;
    }
    if (aState.p) {
      if (typeof process === "object" && process.env.NODE_ENV !== "production") {
        console.warn("Reading pending atom state in write operation. We throw a promise for now.", a);
      }
      throw aState.p;
    }
    if ("v" in aState) {
      return aState.v;
    }
    if (typeof process === "object" && process.env.NODE_ENV !== "production") {
      console.warn("[Bug] no value found while reading atom in write operation. This is probably a bug.", a);
    }
    throw new Error("no value found");
  }, (a, v) => {
    if (a === atom) {
      if (!hasInitialValue(a)) {
        throw new Error("no atom init");
      }
      if (v instanceof Promise) {
        const promise = v.then((resolvedValue) => {
          setAtomValue(state, a, resolvedValue);
          invalidateDependents(state, a);
          flushPending(state);
        }).catch((e) => {
          setAtomReadError(state, atom, e instanceof Error ? e : new Error(e));
          flushPending(state);
        });
        setAtomReadPromise(state, atom, promise);
      } else {
        setAtomValue(state, a, v);
      }
      invalidateDependents(state, a);
    } else {
      writeAtomState(state, a, v);
    }
    flushPending(state);
  }, update);
  if (promiseOrVoid instanceof Promise) {
    const promise = promiseOrVoid.finally(() => {
      setAtomWritePromise(state, atom);
      flushPending(state);
    });
    setAtomWritePromise(state, atom, promise);
  }
};
const writeAtom = (state, writingAtom, update) => {
  writeAtomState(state, writingAtom, update);
  flushPending(state);
};
const isActuallyWritableAtom = (atom) => !!atom.write;
const mountAtom = (state, atom, initialDependent) => {
  const atomState = readAtomState(state, atom);
  atomState.d.forEach((_, a) => {
    if (a !== atom) {
      const aMounted = state.m.get(a);
      if (aMounted) {
        aMounted.d.add(atom);
      } else {
        mountAtom(state, a, atom);
      }
    }
  });
  const mounted = {
    d: new Set(initialDependent && [initialDependent]),
    l: new Set(),
    u: void 0
  };
  state.m.set(atom, mounted);
  if (isActuallyWritableAtom(atom) && atom.onMount) {
    const setAtom = (update) => writeAtom(state, atom, update);
    mounted.u = atom.onMount(setAtom);
  }
  return mounted;
};
const unmountAtom = (state, atom) => {
  var _a;
  const onUnmount = (_a = state.m.get(atom)) == null ? void 0 : _a.u;
  if (onUnmount) {
    onUnmount();
  }
  state.m.delete(atom);
  const atomState = getAtomState(state, atom);
  if (atomState) {
    atomState.d.forEach((_, a) => {
      if (a !== atom) {
        const mounted = state.m.get(a);
        if (mounted) {
          mounted.d.delete(atom);
          if (canUnmountAtom(a, mounted)) {
            unmountAtom(state, a);
          }
        }
      }
    });
  } else if (typeof process === "object" && process.env.NODE_ENV !== "production") {
    console.warn("[Bug] could not find atom state to unmount", atom);
  }
};
const mountDependencies = (state, atom, atomState, prevDependencies) => {
  const dependencies = new Set(atomState.d.keys());
  prevDependencies.forEach((_, a) => {
    if (dependencies.has(a)) {
      dependencies.delete(a);
      return;
    }
    const mounted = state.m.get(a);
    if (mounted) {
      mounted.d.delete(atom);
      if (canUnmountAtom(a, mounted)) {
        unmountAtom(state, a);
      }
    }
  });
  dependencies.forEach((a) => {
    const mounted = state.m.get(a);
    if (mounted) {
      const dependents = mounted.d;
      dependents.add(atom);
    } else {
      mountAtom(state, a, atom);
    }
  });
};
const commitAtomState = (state, atom, atomState, prevDependencies) => {
  if (typeof process === "object" && process.env.NODE_ENV !== "production") {
    Object.freeze(atomState);
  }
  const isNewAtom = !state.a.has(atom);
  state.a.set(atom, atomState);
  if (state.l) {
    state.l(atom, isNewAtom);
  }
  ++state.v;
  if (!state.p.has(atom)) {
    state.p.set(atom, prevDependencies);
  }
};
const flushPending = (state) => {
  const pending = Array.from(state.p);
  state.p.clear();
  pending.forEach(([atom, prevDependencies]) => {
    const atomState = getAtomState(state, atom);
    if (atomState) {
      if (prevDependencies) {
        mountDependencies(state, atom, atomState, prevDependencies);
      }
    } else if (typeof process === "object" && process.env.NODE_ENV !== "production") {
      console.warn("[Bug] atom state not found in flush", atom);
    }
    const mounted = state.m.get(atom);
    mounted == null ? void 0 : mounted.l.forEach((listener) => listener());
  });
};
const restoreAtoms = (state, values) => {
  for (const [atom, value] of values) {
    if (hasInitialValue(atom)) {
      setAtomValue(state, atom, value);
      invalidateDependents(state, atom);
    }
  }
  flushPending(state);
};

const TARGET = Symbol();
const GET_VERSION = Symbol();
const createMutableSource = (target, getVersion) => ({
  [TARGET]: target,
  [GET_VERSION]: getVersion
});

const createStoreForProduction = (initialValues) => {
  const state = createState(initialValues);
  const stateMutableSource = createMutableSource(state, () => state.v);
  const commitCallback = () => flushPending(state);
  const updateAtom = (atom, update) => writeAtom(state, atom, update);
  return [stateMutableSource, updateAtom, commitCallback];
};
const createStoreForDevelopment = (initialValues) => {
  const stateListener = (updatedAtom, isNewAtom) => {
    ++debugStore.version;
    if (isNewAtom) {
      debugStore.atoms = [...debugStore.atoms, updatedAtom];
    }
    Promise.resolve().then(() => {
      debugStore.listeners.forEach((listener) => listener());
    });
  };
  const state = createState(initialValues, stateListener);
  const stateMutableSource = createMutableSource(state, () => state.v);
  const commitCallback = () => flushPending(state);
  const updateAtom = (atom, update) => writeAtom(state, atom, update);
  const debugStore = {
    version: 0,
    atoms: [],
    state,
    listeners: new Set()
  };
  const debugMutableSource = createMutableSource(debugStore, () => debugStore.version);
  const restore = (values) => restoreAtoms(state, values);
  return [
    stateMutableSource,
    updateAtom,
    commitCallback,
    debugMutableSource,
    restore
  ];
};
typeof process === "object" && process.env.NODE_ENV !== "production" ? createStoreForDevelopment : createStoreForProduction;
const isDevStore = (store) => {
  return store.length > 3;
};

const getDebugStateAndAtoms = ({
  atoms,
  state
}) => [state, atoms];
const subscribeDebugStore = ({ listeners }, callback) => {
  listeners.add(callback);
  return () => listeners.delete(callback);
};

function useAtomsSnapshot(scope) {
  const StoreContext = SECRET_INTERNAL_getStoreContext(scope);
  const debugMutableSource = useContext(StoreContext)[3];
  if (debugMutableSource === void 0) {
    throw Error("useAtomsSnapshot can only be used in dev mode.");
  }
  const [state, atoms] = SECRET_INTERNAL_useMutableSource(debugMutableSource, getDebugStateAndAtoms, subscribeDebugStore);
  const atomToAtomValueTuples = atoms.filter((atom) => !!state.m.get(atom)).map((atom) => {
    var _a;
    const atomState = (_a = state.a.get(atom)) != null ? _a : {};
    return [atom, atomState.e || atomState.p || atomState.w || atomState.v];
  });
  return new Map(atomToAtomValueTuples);
}

function useGotoAtomsSnapshot(scope) {
  const StoreContext = SECRET_INTERNAL_getStoreContext(scope);
  const store = useContext(StoreContext);
  if (!isDevStore(store)) {
    throw new Error("useGotoAtomsSnapshot can only be used in dev mode.");
  }
  const restore = store[4];
  return useCallback((values) => {
    for (const [atom] of values) {
      if (atom.scope !== scope) {
        throw new Error("atom scope mismatch to restore");
      }
    }
    restore(values);
  }, [restore, scope]);
}

export { useAtomDevtools, useAtomsSnapshot, useGotoAtomsSnapshot };
